crates/lib/src/execution_data.rs

```diff
@@ -1,120 +1,334 @@
+                            m.column_mut(reference.index)
+                                .assign(&Array::from_vec(v.clone()));
+                        }
+                        (RegisterMatrix::Real(m), ReadoutValues::Real(v))
+                            if m.nrows() == v.len() =>
+                        {
+                            m.column_mut(reference.index)
+                                .assign(&Array::from_vec(v.clone()));
+                        }
+                        (RegisterMatrix::Complex(m), ReadoutValues::Complex(v))
+                            if m.nrows() == v.len() =>
+                        {
+                            m.column_mut(reference.index)
+                                .assign(&Array::from_vec(v.clone()));
+                        }
+                        _ => {
+                            return Err(RegisterMatrixConversionError::InvalidShape {
+                                register: reference.name,
+                            })
+                        }
+                    }
+                    Ok(register_map)
+                },
+            )?,
+        ))
     }
 }
 
-impl From<HashMap<MemoryReference, ReadoutValues>> for ReadoutMap {
-    fn from(map: HashMap<MemoryReference, ReadoutValues>) -> Self {
-        Self(map)
-    }
-}
-
-/// The result of executing an [`Executable`](crate::Executable) via
-/// [`Executable::execute_on_qpu`](crate::Executable::execute_on_qpu).
-#[derive(Debug, Clone, PartialEq)]
-pub struct Qpu {
-    /// The data of all readout data that were read from
-    /// (via [`Executable::read_from`](crate::Executable::read_from)). Key is the name of the
-    /// register, value is the data of the register after execution.
-    pub readout_data: ReadoutMap,
-    /// The time it took to execute the program on the QPU, not including any network or queueing
-    /// time. If paying for on-demand execution, this is the amount you will be billed for.
-    ///
-    /// This will always be `None` for QVM execution.
-    pub duration: Option<Duration>,
+// This is a copy of [`quil_rs::instruction::MemoryReference`] that uses `usize` for the index
+// instead of `u64` for compatibility with the containers we use for [`RegisterMap`].
+// It's possible `quil_rs` will use `usize` for its `MemoryReference` in the future. If so, we
+// should use it to replace this.
+// See https://github.com/rigetti/qcs-sdk-rust/issues/224
+#[derive(Debug, PartialEq, PartialOrd, Eq, Ord)]
+struct MemoryReference {
+    name: String,
+    index: usize,
 }
 
 #[derive(Debug, thiserror::Error)]
-pub(crate) enum MemoryReferenceParseError {
+pub enum MemoryReferenceParseError {
     #[error("Could not parse memory reference: {reason}")]
     InvalidFormat { reason: String },
 
     #[error("Could not parse index from reference name: {0}")]
     InvalidIndex(#[from] ParseIntError),
 }
 
-// Note: MemoryReference may have a from_string in the fututre
 fn parse_readout_register(
```

Comment:
this _should_ be [this](https://github.com/rigetti/quil-rs/blob/ba5f28987a1c64efbde1c621ec219e7d85f7128c/src/instruction.rs#L340) - no custom function or error type needed.

Reply:
It appears the future is here.

I updated the function to use `from_str`, and the error type to match, but I ultimately kept them, because it's a useful abstraction for the readout data to register map conversion. We need the `index` of a memory reference to be `usize` since it is used to initialize matrices, and catching that here instead of every other place it's used.

Reply:
Would it still make sense to use the official parser and then convert from one MemoryReference to the other? Unlikely to make a difference in practice, but that saves us two possibly-differing parser implementations. One difference even in this simple parser is that `ro` is a valid memory reference but, lacking brackets, would not parse correctly here.